File: 0079-word-search/0079-word-search.cpp
class Solution {
public:

    bool found_next_char(vector<vector<char> >& board, string& word, int x, int y, int index)
    {
		char	tmp = board[y][x];
		++index;
        if (index == word.size())
            return true;
		board[y][x] = 0;
		if (index == 2)
			;
        if (y - 1 >= 0 && board[y - 1][x] == word[index] && found_next_char(board, word, x, y - 1, index))
            return true;
        else if (x + 1 < board[y].size() && board[y][x + 1] == word[index] && found_next_char(board, word, x + 1, y, index))
            return true;
        else if (y + 1 < board.size() && board[y + 1][x] == word[index] && found_next_char(board, word, x, y + 1, index))
            return true;
        else if (x - 1 >= 0 && board[y][x - 1] == word[index] && found_next_char(board, word, x - 1, y, index))
            return true;
		board[y][x] = tmp;
        return false;
    }

    bool exist(vector<vector<char> >& board, string word) {
        int index = 0;

        for (int i = 0, m = board.size(); i < m; ++i)
        {
            for(int j = 0, n = board[0].size(); j < n; ++j)
            {
                if(board[i][j] == word[0] && found_next_char(board, word, j, i, index))
                    return true;
            }
        }
        return false;
    }
};

File: 0079-word-search/0079-word-search.java
class Solution {
    int x_size, y_size;
    boolean[][] visit;
    int[] dirX = {0, 1, 0, -1};
    int[] dirY = {1, 0, -1, 0};
    public boolean exist(char[][] board, String word) {
        y_size = board.length;
        x_size = board[0].length;
        visit = new boolean[y_size][x_size];
        char[] charArr = word.toCharArray();

        for(int i = 0; i < y_size; ++i) {
            for(int j = 0; j < x_size; ++j) {
                if(dfs(i,j,0,board,charArr))
                    return true;
            }
        }
        return false;
    }

    boolean dfs(int y, int x, int dep, char[][] board, char[] charArr) {
        if(dep == charArr.length)
            return true;
        if(y < 0 || x < 0 || y >= y_size || x >= x_size)
            return false;
        if(visit[y][x] == true || board[y][x] != charArr[dep])
            return false;
        visit[y][x] = true;
        for(int i = 0; i < 4; ++i) {
            if(dfs(y + dirY[i], x + dirX[i], dep + 1, board, charArr))
                return true;
        }
        visit[y][x] = false;
        return false;
    }
}